exporters/prometheus/src/exporter_utils.cc

```diff
@@ -260,22 +361,118 @@ void PrometheusExporterUtils::SetData(std::vector<T> values,
+      if (label_name == kPrometheusInstance)
+      {
+        has_instance_label = true;
+      }
+      else if (label_name == kPrometheusJob)
+      {
+        has_job_label = true;
+      }
+      AddPrometheusLabel(std::move(label_name), AttributeValueToString(label.second),
+                         &metric.label);
     }
   }
+
+  // Convert resource to prometheus labels
+  if (nullptr != resource)
+  {
+    do
+    {
+      if (has_job_label)
+      {
+        break;
+      }
+
+      opentelemetry::sdk::resource::ResourceAttributes::const_iterator prometheus_job_it =
+          resource->GetAttributes().find(kPrometheusJob);
+      if (prometheus_job_it != resource->GetAttributes().end())
+      {
+        AddPrometheusLabel(kPrometheusJob, AttributeValueToString(prometheus_job_it->second),
+                           &metric.label);
+        break;
+      }
+
+      opentelemetry::sdk::resource::ResourceAttributes::const_iterator service_name_it =
+          resource->GetAttributes().find(
+              opentelemetry::sdk::resource::SemanticConventions::kServiceName);
+      opentelemetry::sdk::resource::ResourceAttributes::const_iterator service_namespace_it =
+          resource->GetAttributes().find(
+              opentelemetry::sdk::resource::SemanticConventions::kServiceNamespace);
+
+      if (service_namespace_it != resource->GetAttributes().end() &&
+          service_name_it != resource->GetAttributes().end())
+      {
+        AddPrometheusLabel(kPrometheusJob,
+                           AttributeValueToString(service_namespace_it->second) + "/" +
+                               AttributeValueToString(service_name_it->second),
+                           &metric.label);
+      }
+      else if (service_name_it != resource->GetAttributes().end())
+      {
+        AddPrometheusLabel(kPrometheusJob, AttributeValueToString(service_name_it->second),
+                           &metric.label);
+      }
+      else if (service_namespace_it != resource->GetAttributes().end())
+      {
+        AddPrometheusLabel(kPrometheusJob,
+                           AttributeValueToString(service_namespace_it->second) + "/",
+                           &metric.label);
+      }
+    } while (false);
+
+    if (!has_instance_label)
+    {
+      opentelemetry::sdk::resource::ResourceAttributes::const_iterator service_instance_id_it =
+          resource->GetAttributes().find(
+              opentelemetry::sdk::resource::SemanticConventions::kServiceInstanceId);
+      if (service_instance_id_it == resource->GetAttributes().end())
+      {
+        service_instance_id_it = resource->GetAttributes().find(kPrometheusInstance);
+      }
+      if (service_instance_id_it != resource->GetAttributes().end())
+      {
+        has_instance_label = true;
+        AddPrometheusLabel(kPrometheusInstance,
+                           AttributeValueToString(service_instance_id_it->second), &metric.label);
+      }
+    }
+  }
+
```

Comment:
Can you please revisit this function - looks like we are somewhat overengineering stuff here. The `job` and `instance` labels are supposed to be handled in the Prometheus exporter implemented for Collector, and it is already done here - https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusexporter/collector.go.

Reply:
I found no description about how to export labels when users already set `job` and `instance` as attributes. In my opinion, we can keep compatibility with old codes which already set `job` and `instance` as attributes of metrics data or resource.
Or do you think we can keep one or two versions to let users use old ways?

Reply:
> I found no description about how to export labels when users already set job and instance as attributes.

Yes, as the specs don't specify any mapping for `job` and `instance` set within attributes/resources, we shouldn't be adding them. And let Prometheus backend add this data itself, as it has the endpoint URL it is scrapping to, as specified here - https://prometheus.io/docs/concepts/jobs_instances/. Also, needn't ignore any of the resources/attributes while copying them to target_info/labels.

The specs define a mapping for the Collector's Prometheus exporter, as this exporter could be fetching data from multiple targets, so each metric data could have a different `job` and `instance` value depending on the target.

>  Or do you think we can keep one or two versions to let users use old ways?

We can keep the existing version i.e., sanitize and copy all the metrics attributes to Prometheus labels? No need to ignore resources or attributes during conversions.

I am talking on the basis of reading specs - do you see any issue with that from an end-user perspective?

Reply:
I mean if a metric has resource attribute `service.instance.id` and attribute `instance`. If we do not ignore one, there will be two `instance` in prometheus metric.The implementation in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusexporter/collector.go use `map[string]string` to set labels from metric attributes and then from resource attributes, which means if when both `service.instance.id` and `instance` are set, it will use `service.instance.id`. But in prometheus-cpp, it use `std::vector<Label>`, not a `map, if we do not ignore one, there will be two `instance` labels.
I just update the code to keep the same behaviour as it in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusexporter/collector.go . The `job` and `[service.namespace/]service.name` have the same problem.

Reply:
SDK exporters (not collector) should simply map all resource attributes to labels on the target_info metric with no filtering or translation.  I realize there can possibly be collisions if someone is using prometheus exporters in the collector, but that is best solved by receiver config (honor_labels) or processor (e.g. resource) configuration.

Reply:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusexporter/collector.go#L302
This collector also remove these attributes. Should we keep it the same behaviour as the collector?

Reply:
We should not.  The collector and SDK exporters intentionally have different behavior.  That behavior (converting service.name, service.namespace, and service.instance.id to job and instance) is undoing what is done in the prometheus receiver.  Since this (and other SDK) exporter does not receive metrics from other prometheus sources, it shouldn't include that logic.